text/src/autogluon/text/automm/data/infer_types.py

```diff
@@ -0,0 +1,301 @@
+import collections
+import pandas as pd
+import warnings
+from typing import Union, Optional, List, Dict, Tuple
+from pandas import DataFrame, Series
+from ..constants import NULL, CATEGORICAL, NUMERICAL, TEXT, IMAGE_PATH, MULTICLASS, BINARY, REGRESSION
+
+
+def is_categorical_column(data: pd.Series,
+                          valid_data: pd.Series,
+                          threshold: int = None,
+                          ratio: Optional[float] = None,
+                          oov_ratio_threshold: Optional[float] = None,
+                          is_label: bool = False,
+                          ) -> bool:
+    """Check whether the column is a categorical column.
+
+    If the number of unique elements in the column is smaller than
+
+        min(#Total Sample * ratio, threshold),
+
+    it will be treated as a categorical column.
+
+    Parameters
+    ----------
+    data
+        The column data
+    valid_data
+        Additional validation data
+    threshold
+        The threshold for detecting categorical column
+    ratio
+        The ratio detecting categorical column
+    oov_ratio_threshold
+        The out-of-vocabulary ratio between training and validation.
+        This is used to determine if the column is a categorical column.
+        Usually, a categorical column can tolerate a small OOV ratio
+    is_label
+        Whether the column is a label column.
+
+    Returns
+    -------
+    is_categorical
+        Whether the column is a categorical column
+    """
+    if data.dtype.name == 'category':
+        return True
+    else:
+        if threshold is None:
+            if is_label:
+                threshold = 100
+                oov_ratio_threshold = 0
+                ratio = 0.1
+            else:
+                threshold = 20
+                oov_ratio_threshold = 0
+                ratio = 0.1
+        threshold = min(int(len(data) * ratio), threshold)
+        data_value_counts = data.value_counts(dropna=False)
+        key_set = set(data_value_counts.keys())
+        if len(data_value_counts) < threshold:
+            valid_value_counts = valid_data.value_counts(dropna=False)
+            total_valid_num = len(valid_data)
+            oov_num = 0
+            for k, v in zip(valid_value_counts.keys(), valid_value_counts.values):
+                if k not in key_set:
+                    oov_num += v
+            if is_label and oov_num != 0:
+                return False
```

Comment:
If we have `[1, 2]` in training label and `[1, 2, 3]` in validation set, why this method would return this is not categorical variable? Granted that it's a bad news if we got into this situation, but it is totally possible if the splits are generated using methods other than stratification by the label. 

If this one is not categorical (assuming we have only a few values on a large dataset), that would make it a regression problem?

Reply:
We can fix the logic after aligning the API design for inferring types.